#167573194 Build Out User Signup Feature

chialuka opened the pull request:
#### What does this PR do?
- This PR builds out the user signup endpoint to enable registering users to use the application

#### Description of Task proposed in this pull request?
- Models, Controllers, and Routes were created to enable creating new users

- Helper functions were created for delivering responses to users and hashing passwords

- Tests were written for the feature

#### How should this be manually tested (Quality Assurance)?
- Clone the repo and run `npm test`

- Aside from this, make a `POST` request to the `/api/v1/users` endpoint and provide these user details using Postman, Insomnia or the Swagger documentation:
`{`
    `"firstName": "Adeola"`
    `"lastName": "Afolabi"`
    `"userName": "ade"`
    `"email": "ade@gmail.com"`
    `"password": "incorrect"`
    `"confirmPassword": "incorrect"`
}`

. The user will get a response with successful registration details

#### What are the relevant pivotal tracker stories?
- [#167573194](https://www.pivotaltracker.com/story/show/167573194)

#### Any background context you want to add (Operations Impact)?
- `npm install`

- Add `BASE_URL` to .env file in the root directory

#### What I have learned working on this feature: 
- I learned how to run Sequelize models and migrations. I also got better at how to mock API calls

#### Screenshots: 
<img width="1440" alt="Screenshot 2019-07-30 at 6 33 20 PM" src="https://user-images.githubusercontent.com/42539315/62153231-0791a900-b2fc-11e9-9528-7c40195ed5f1.png">

henryade commented on server/controllers/Users.js:
```diff
@@ -0,0 +1,47 @@
+
+/**
+ * @export
+ * @class User
```
this should be `Users`

chialuka replied:
> Why is there no confirm password field?

We asked the PO for the fields to use and she said password field is okay. There was no confirm password field in the fields she approved of. I followed that. Moreover, the confirm password field can be handled from the frontend and in validations

henryade commented on server/controllers/Users.js:
```diff
@@ -0,0 +1,47 @@
+   * @name signUp
+   * @async
+   * @static
+   * @memberof User
```
this should be Users also

henryade reviewed (commented):
Why is there no confirm password field?

Adekoreday commented on server/database/seeds/20190730180408-my-seed-file.js:
```diff
@@ -0,0 +1,19 @@
+module.exports = {
+  up: (queryInterface, Sequelize) => queryInterface.bulkInsert(
```
kindly fix this

chialuka replied:
Thank you. I'll do that right away

Adekoreday reviewed (commented):
kindly fix un-used vars

ayodejiAA commented on server/controllers/Users.js:
```diff
@@ -0,0 +1,47 @@
+      // const token = await generateToken(user.id);
+      return serverResponse(res, 201, { user });
+    } catch (error) {
+      return serverError(res);
```
Nice work. I like the fact that you also cater for error handling. However, don't you think the server error response should be called directly in every controller method? Why don't we have an express error handler middleware in our server entry file which is in place to respond also to the errors that may occur even when controller method connection is closed.

chialuka committed:
feature-[167573194]: build out signup endpoint
- Enable user sign up feature
- Write unit and integration tests for feature
[Deivers #167573194]

chialuka committed:
feature-[167573194]: build out signup endpoint
- Enable user sign up feature
- Write unit and integration tests for feature
[Deivers #167573194]

dinobi commented on server/controllers/Users.js:
```diff
@@ -0,0 +1,52 @@
+        return serverResponse(res, 400, { error: 'Passwords do not match' });
+      }
+      if (await checkEmail(req.body.email)) {
+        return serverResponse(res, 409, { error: 'Email already in use' });
```
change to `Email has already been taken`

chialuka replied:
Ok. I'll do that right away

dinobi commented on server/controllers/Users.js:
```diff
@@ -0,0 +1,52 @@
+        return serverResponse(res, 409, { error: 'Email already in use' });
+      }
+      if (await checkUserName(req.body.userName)) {
+        return serverResponse(res, 409, { error: 'Username already in use' });
```
same here

dinobi commented on server/routes/user.js:
```diff
@@ -0,0 +1,8 @@
+
+const route = express.Router();
+
+route.post('/create', Users.signUp);
```
pls change to: `route.post('/users/create', Users.create)`
This means you have to change the method name.
Make this api change also reflect on the README.md

chialuka replied:
Noted

dinobi commented on server/routes/index.js:
```diff
@@ -0,0 +1,8 @@
+
+const route = express.Router();
+
+route.use('/users', users);
```
what is `users`?, I've seen this somewhere and I'm asking again

chialuka replied:
Users is a route. All routes that contain `/user/` will be routed to the `routes/users/` file. Doing it this way will allow us abstract the different route files for users, comments, articles etc

dinobi commented on server/helpers/serverResponse.js:
```diff
@@ -0,0 +1,19 @@
+ * @returns {JSON} JSON response with server error details
+ */
+const serverError = res => res.status(500).json({
+  error: 'A fix is in progress'
```
can we change to `server error, this will be resolved shortly`, all small case

chialuka replied:
Yes. That is done

dinobi commented on package.json:
```diff
@@ -16,7 +16,7 @@
     "start": "node dist/index.js",
     "start:dev": "DEBUG=dev nodemon --exec babel-node server",
-    "test": "nyc mocha --timeout 1000 -r esm --exit"
+    "test": "npm run db:rollback && nyc mocha --timeout 1000 -r esm --exit"
```
I wonder why we still have this again here

chialuka replied:
It is for running migrations before the test runs. If we don't have it then our tests will fail on Travis because there is no migration to run. Remember I had to remove it from the bug fix branch because I also removed the migrations and models.

dinobi commented on server/controllers/Users.js:
```diff
@@ -0,0 +1,56 @@
+   * @param {Object} res express response object
+   * @returns {JSON} JSON object with details of new user
+   */
+  static async signUp(req, res) {
```
change to `create`

chialuka replied:
Ok. Noted

dinobi commented on server/controllers/Users.js:
```diff
@@ -0,0 +1,56 @@
+        password: hashedPassword
+      });
+      delete user.dataValues.password;
+      // const token = await generateToken(user.id);
```
Pls remove commented code

chialuka replied:
Okay. It was added to make provision for the token to be sent to the user. I'll remove them innediately

dinobi commented on server/database/models/user.js:
```diff
@@ -0,0 +1,17 @@
+    {
+      firstName: DataTypes.STRING,
+      lastName: DataTypes.STRING,
+      userName: DataTypes.STRING,
```
I don't see the uniqueness constraint for username and email, how did you define the uniqueness?

chialuka replied:
Uniqueness and non-nullable values are defined in the migrations

chialuka committed:
Merge branch 'develop' into ft-user-signup-167573194

chialuka committed:
Update Users.js
